labellab-flask/api/controllers/issuecontroller.py

```diff
@@ -3,36 +3,136 @@
 from flask import make_response, request, jsonify, current_app
 from flask_jwt_extended import (
     jwt_required,
+    get_jwt_identity,
 )
 
 from api.config import config
 from api.models.Issue import Issue
 
+from api.helpers.issue import (
+    save as save_issue
+)
+from api.middleware.project_member_access import project_member_only
+
+allowed_priorities = config[os.getenv("FLASK_CONFIG") or "development"].ISSUE_PRIORITIES_ALLOWED
+allowed_statuses = config[os.getenv("FLASK_CONFIG") or "development"].ISSUE_STATUSES_ALLOWED
+allowed_categories = config[os.getenv("FLASK_CONFIG") or "development"].CATEGORIES_ALLOWED
+allowed_entity_types = config[os.getenv("FLASK_CONFIG") or "development"].ENTITY_TYPES_ALLOWED
+
 class CreateIssues(MethodView):
-    """
-    This class-based view handles fetching all issues within a project
-    Url --> /api/v1/issue/create/<int:project_id>
-    """
+    """This class creates a new Issue."""
     @jwt_required
+    @project_member_only
     def post(self, project_id):
+        """
+        Handle POST request for this view.
+        Url --> /api/v1/issue/create/<int:project_id>
+        """
+        # getting JSON data from request
+        post_data = request.get_json(silent=True, force=True)
+        current_user = get_jwt_identity()
+        # Load model with necessary fields 
         try:
-                        
+            title = post_data["title"]
+            description = post_data["description"]
+            category = post_data["category"]
+        except Exception:
             response = {
-                "success": True,
-                "msg": "Issue Created",
-                "body": "New Issue Create for this project"
+                "success": False,
+                "msg": "Please provide all the required fields."
             }
-            return make_response(jsonify(response)), 200
+            return make_response(jsonify(response)), 400
+        if category not in allowed_categories:
+                print("Error occured: category not allowed")
+                response = {
+                        "success": False,
+                        "msg": "category not allowed."
+                    }
+                return make_response(jsonify(response)), 400
         
-        except Exception:
+        """Save the new Issue"""
+        try:
+            issue = Issue(
+                title=title, 
+                description=description, 
+                project_id=project_id,
+                created_by=current_user,
+                category=category,
+            )
+            # Save the model with optional fields
```

Comment:
We're writing a lot of repeated code here for the optional fields. Let's try to minimize it. We can create a util for that. One such approach can be -

1. We can create a list of dictionaries in the helpers file. Each dict would contain info about an optional attribute.

```
issue_attribute_validator = [
    {
        'key': 'priority',
        'enum': allowed_priorities,
    },
    {
        'key': 'entity_id',
    }
]
```

`key` refers to name of the attribute. `enum` can be an optional key in the dict referring to what values does that field allow.

2. And then inside the controller, we can loop over the list and do the check there itself.

```
for attribute in issue_attribute_validator:
    try:
        attribute_value = post_data[attribute['key']]

        # Check if enum is there
        # If yes, validate against it
        if 'enum' in attribute and attribute_value not in attribute['enum']:
            response = {
                'success': False,
                'msg': f'{attribute["key"]} has an invalid value'
            }
            return make_response(jsonify(response)), 400

        # Set attribute in issue object
        setattr(issue, attribute['key'], attribute_value)
    except:
        # If the field is not present then just move on
        continue
```

Reply:
This is just something rough I wrote for your reference. Do let me know if you need my help with this.